ext/standard/proc_open.c

```diff
@@ -943,6 +943,268 @@ static int set_proc_descriptor_from_resource(zval *resource, descriptorspec_item
 	return SUCCESS;
 }
 
+#ifdef PHP_WIN32
+
+#define THREADED_PIPE_BUF_SIZE 4096
+
+typedef struct _threaded_pipe {
+	HANDLE pipe;         /* Parent end of a process pipe. */
+	HANDLE thread;       /* Thread being used to handle IO. */
+	php_socket_t sock;   /* Internal end of the proxy socket pair. */
+	php_stream *stream;  /* Wraps the exposed end of the socket pair. */
+	const char *mode;    /* Pipe open mode. */
+	int closed;          /* Set to 1 on stream close. */
+	int wsa_error;       /* Last socket error code. */
+	DWORD error;         /* Last error code. */
+} threaded_pipe;
+
+static int threaded_error_check(threaded_pipe *data, const char *format)
+{
+	if (data->error) {
+		char *err = php_win32_error_to_msg(data->error);
+		php_error_docref(NULL, E_WARNING, format, (int) data->error, err);
+		php_win32_error_msg_free(err);
+		return 1;
+	}
+
+	if (data->wsa_error) {
+		zend_string *err = php_socket_error_str(data->wsa_error);
+		php_error_docref(NULL, E_WARNING, format, data->wsa_error, ZSTR_VAL(err));
+		zend_string_release(err);
+		return 1;
+	}
+
+	return 0;
+}
+
+static ssize_t threaded_pipe_write(php_stream *stream, const char *buf, size_t count)
+{
+	threaded_pipe *data = (threaded_pipe *) stream->abstract;
+
+	if (threaded_error_check(data, "Pipe write failed: [%d] %s")) {
```

Comment:
I would prefer passing in only `"Pipe write failed"` and let the rest of the message be constructed by `threaded_error_check`. Separating format string from arguments makes me uncomfortable...

Reply:
I also don't really get why the error check is performed *before* the operation?